Finite HMM api

vruano opened the pull request:
HMM api.
@davidbenjamin  please review.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+ *
+ * <p>This done by calling {@link #apply} that returns an
+ * {@link ForwardBackwardAlgorithm.Result} instance which
+ * allow for subsequent queries on the posterior probability of the
```
Nice design decision.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+    /**
+     * The forward-backward algorithm result query interface.
+     *
+     * @param <D> the data observation type.
```
"observed data type" reads better.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/HiddenMarkovModel.java:
```diff
@@ -0,0 +1,108 @@
+ * emission that might depends on the position.
+ *
+ * <p>
+ * This class models a process that emits a fix length data sequence.
```
fixed-length

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/HiddenMarkovModel.java:
```diff
@@ -0,0 +1,108 @@
+     * <p>
+     *     Contract:
+     *     <ul>
+     *         <li>This method must always return the same state every time is called.</li>
```
same *states* every time *it* is called.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+            throw new IllegalArgumentException("the data sequence and position sequence must have the same number of elements");
+        }
+        final List<S> states = model.hiddenStates();
+        Utils.nonNull(states, "the model hidden state list cannot be null");
```
Shouldn't `HiddenMarkovModel` itself (or its implementations) enforce this?

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+        }
+        final List<S> states = model.hiddenStates();
+        Utils.nonNull(states, "the model hidden state list cannot be null");
+        if (states.size() == 0) {
```
Likewise.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+     * @throws IllegalArgumentException if the {@code model} does not recognize any of the values in {@code dataList}
+     *   or {@code positionList}.
+     */
+    private static <D, T, S> double[] calculateLogBackwardProbabilities(final HiddenMarkovModel<D, T, S> model,
```
Why not `double[][]`?

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+     * @throws IllegalArgumentException if the {@code model} does not recognize any of the values in {@code data}
+     *   or {@code positions}.
+     */
+    private static <D, T, S> double[] calculateLogForwardProbabilities(
```
Forward probabilities should go above backward probabilities in the code.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+            return result;
+        }
+
+        // We first initialize the probabilities at the first position.
```
This comment should be very explicit about the fact that we initialize the last position of the chain, which is the first position of the backward pass.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+        }
+
+        // We first initialize the probabilities at the first position.
+        // result already contains appropriate values for the backward-probabilities
```
"Result already contains" implies that this follows from a calculation.    Better just to say we initialize to 1.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+                                                                   final List<T> positionList) {
+
+        final List<S> states = model.hiddenStates();
+        final int width = states.size();
```
`numStates` would be clearer.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+
+        // We first initialize the probabilities at the first position.
+        // result already contains appropriate values for the backward-probabilities
+        // of the last state = 1, (0 in log scale).
```
Also, make it explicit that declaring the `double[]` automatically initializes to zero.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+        // of the last state = 1, (0 in log scale).
+
+        final double[] logSumBuffer = new double[states.size()];
+        for (int t = length - 2; t >= 0; --t) {
```
`t` and `u` need clearer names, like `thisPosition` and `nextPosition` or `here` and `next` etc.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+            final int uOffset = tOffset + width;
+            final T positionT = positionList.get(t);
+            final T positionU = positionList.get(u);
+            for (int sT = 0; sT < width; sT++) {
```
Similarly, more evocative names like `thisState` and `nextState` etc.  (Not saying that these particular suggestions are good, but you get the idea).

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+            final List<T> positions) {
+
+        final List<S> states = model.hiddenStates();
+        final int width = states.size();
```
Similar comments to those in backward probability code about variable names.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+         *     be the same.
+         * </p>
+         * <p>
+         *     This implementation uses always the first position.
```
always uses

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+        private final Object2IntMap<T> positionIndex;
+        private final Object2IntMap<S> stateIndex;
+
+        private final int width;
```
`numStates`

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java:
```diff
@@ -0,0 +1,432 @@
+            if (logForwardProbabilities.length == 0) {
+                return 0;
+            } else {
+                final double[] values = new double[width];
```
If probabilities are 2-D arrays this code becomes `return logSumLog(arraySum(forward[0], backward[0]))`.

vruano replied:
It might sound like premature optimization, but I rather don't have ~150,000 3-position arrays.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/HiddenMarkovModel.java:
```diff
@@ -0,0 +1,108 @@
+ * of hidden states which are shared by all positions.
+ * </p>
+ * <p>
+ * The emission probability of a datum given the hidden state may depend on the current hidden state itself and also the position in the sequence.
```
omit "itself"

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ViterbiAlgorithm.java:
```diff
@@ -0,0 +1,178 @@
+public final class ViterbiAlgorithm {
+
+    /**
+     * Calculates the most likely the hidden state sequence that explains the observed data
```
extra "the"

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ViterbiAlgorithm.java:
```diff
@@ -0,0 +1,178 @@
+                // Initialize best-t-state search setting the best so
+                // far to 0th indexed state:
+                int bestTStateIndex = 0;
+                double bestTStateLog10Prob = bestTPaths[0].log10Prob
```
There is a mix of `log` and `log10` variable names.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hmm/ViterbiAlgorithm.java:
```diff
@@ -0,0 +1,178 @@
+        };
+
+        for (int u = 1; u < length; u++) {
+            final int t = u - 1;
```
Same stuff about variable names.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/utils/hmm/HMMUnitTest.java:
```diff
@@ -0,0 +1,542 @@
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public final class HMMUnitTest extends BaseTest {
```
I'm uncomfortable about relying so heavily on R for these tests.  You should supplement it with a few HMMs that are a priori trivial.  For example:

* If priors and transitions are all equal, then every path must be equally likely.
* If one state has the likeliest prior and most likely transitions from any state are into that state, then the likeliest path is that state repeated over and over.
* If you have a long chain with (irreducible) transition matrix M, the state probabilities asymptotically approach the eigenvector with largest eigenvalue of M.  (I consider this test very important because it would be basically impossible to get the right answer by coincidence).

davidbenjamin replied:
A simpler and perfectly adequate version of that last one is that if the initial probabilities vector is v and the (position-dependent) transition matrices are M_1, M_2. . . then the probabilities vector at position k is M_k M_{k-1} . . . M_2 M_1 v.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/utils/hmm/TestHMModel.java:
```diff
@@ -0,0 +1,232 @@
+                return Datum.values()[i];
+            }
+        }
+        throw new IllegalStateException("may be some of the emission probs is negative?");
```
are negative

davidbenjamin commented:
@vruano Back to you.  It's a nice design, very clear API.  My main suggestion is to supplement the R testing with a few analytic results.  Also, I wasn't fully comfortable with the performance hit  representing states by objects instead of integer indices, but then I saw your testing code's use of `ordinal()`  I can't say I love this solution, but I don't know of any better way and I do like the flexibility of using either an `Integer` or an `Enum`.

vruano commented:
@davidbenjamin you can always use Integer typed hidden-states if that is what you need.

vruano commented:
@davidbenjamin  back to you. 
I have addressed all the comments in three commits 
196acc5 : simple comments
271b82f : HMM enhanced unit testing.
7413b6a : FW tables from 1D to 2D arrays.

davidbenjamin commented:
Very small comments.  I love the HMM tests.  Feel free to merge when you're done.
